rsocket/RSocket.cpp

```diff
@@ -2,77 +2,106 @@
+          ConnectionFactory::ConnectedDuplexConnection connection) mutable {
+        // fromConnection method must be called from the eventBase and since
+        // there is no guarantee that the Future returned from the
+        // connectionFactory::connect method is executed on the event base,
+        // we have to ensure it by using folly::via
+        auto* eventBase = &connection.eventBase;
+        return via(eventBase).then(
+            [connection = std::move(
+                connection), createRSC = std::move(createRSC)]() mutable {
+              return createRSC(std::move(connection));
+            });
+      });
 }
 
-folly::Future<std::shared_ptr<RSocketClient>> RSocket::createResumedClient(
-    std::unique_ptr<ConnectionFactory> connectionFactory,
-    SetupParameters setupParameters,
+folly::Future<std::unique_ptr<RSocketClient>> RSocket::createResumedClient(
+    std::shared_ptr<ConnectionFactory> connectionFactory,
+    ResumeIdentificationToken token,
     std::shared_ptr<ResumeManager> resumeManager,
     std::shared_ptr<ColdResumeHandler> coldResumeHandler,
-    OnRSocketResume,
     std::shared_ptr<RSocketResponder> responder,
     std::unique_ptr<KeepaliveTimer> keepaliveTimer,
     std::shared_ptr<RSocketStats> stats,
-    std::shared_ptr<RSocketConnectionEvents> connectionEvents) {
-  auto c = std::shared_ptr<RSocketClient>(new RSocketClient(
+    std::shared_ptr<RSocketConnectionEvents> connectionEvents,
+    ProtocolVersion protocolVersion) {
+  auto* c = new RSocketClient(
       std::move(connectionFactory),
-      std::move(setupParameters),
+      std::move(protocolVersion),
+      std::move(token),
       std::move(responder),
       std::move(keepaliveTimer),
       std::move(stats),
       std::move(connectionEvents),
       std::move(resumeManager),
-      std::move(coldResumeHandler)));
+      std::move(coldResumeHandler));
 
-  return c->resume().then([c]() mutable { return c; });
+  return c->resume().then(
+      [client = std::unique_ptr<RSocketClient>(c)]() mutable {
+        return std::move(client);
+      });
 }
 
-std::shared_ptr<RSocketClient> RSocket::createClientFromConnection(
+std::unique_ptr<RSocketClient> RSocket::createClientFromConnection(
     std::unique_ptr<DuplexConnection> connection,
     folly::EventBase& eventBase,
     SetupParameters setupParameters,
+    std::shared_ptr<ConnectionFactory> connectionFactory,
     std::shared_ptr<RSocketResponder> responder,
     std::unique_ptr<KeepaliveTimer> keepaliveTimer,
     std::shared_ptr<RSocketStats> stats,
     std::shared_ptr<RSocketConnectionEvents> connectionEvents,
     std::shared_ptr<ResumeManager> resumeManager,
-    std::shared_ptr<ColdResumeHandler> coldResumeHandler,
-    OnRSocketResume) {
-  auto c = std::shared_ptr<RSocketClient>(new RSocketClient(
-      nullptr,
-      std::move(setupParameters),
+    std::shared_ptr<ColdResumeHandler> coldResumeHandler) {
+  auto c = std::unique_ptr<RSocketClient>(new RSocketClient(
```

Comment:
Nit: `std::make_unique<RSocketClient>(...)`

Reply:
ctor of RSocketClient is private..